### PR TITLE
Update v1.Taint parser to accept the form `key:effect` and `key=:effect-`

### LINKED ISSUE #73249: Taint-to-string method generates string that taint parser considers invalid
**What happened**: The [`ToString` receiver on the `v1.Taint` type generates the string](https://github.com/kubernetes/kubernetes/blob/33b20f2/pkg/apis/core/taint.go#L33) "key:effect" when the `Value` of the taint is the empty string.

Because the generated string has no `=` character, when it is passed to the `parseTaint` method in `pkg/util/taints`, the method returns an error.

I discovered this issue by defining a taint in a kubeadm configuration file. kubeadm converted the taint to a string, then passed the string to kubelet using the `--register-with-taints` flag, and kubelet threw an error.

**What you expected to happen**:
I expect the `ToString` receiver on the `v1.Taint` type to only generate strings that are accepted by the `parseTaint` method in `pkg/util/taints`. Conversely, I expect the `parseTaint` method to accept any string generated by the `ToString` receiver on the `v1.Taint` type.

**How to reproduce it (as minimally and precisely as possible)**:
The following program demonstrates the issue. I compiled it using the `release-1.13` branch. It initializes a `Taint` struct with the `Value` field left as an empty string. It then converts the struct to a string. Finally, it calls `utils.NewTaintsVar`, just as kubelet does [here](https://github.com/kubernetes/kubernetes/blob/release-1.13/cmd/kubelet/app/options/options.go#L406). Subsequently, [a call to `Set` ](https://github.com/kubernetes/kubernetes/blob/5fdad5c/pkg/util/taints/taints.go#L85) returns an error that originates in the `parseTaint` method.
```
package main

import (
	"fmt"

	"k8s.io/api/core/v1"
	"k8s.io/kubernetes/pkg/apis/core"
	utiltaints "k8s.io/kubernetes/pkg/util/taints"
)

func main() {
	t := v1.Taint{
		Key:    "node-role.kubernetes.io/master",
		Effect: v1.TaintEffectNoSchedule,
	}
	s := t.ToString()
	ptr := []core.Taint{}
	tv := utiltaints.NewTaintsVar(&ptr)
	if err := tv.Set(s); err != nil {
		fmt.Println(err)
	}
}

```
Output:
```
invalid taint spec: node-role.kubernetes.io/master:NoSchedule
```

### PR DESCRIPTION
**Pull request by dlipovetsky**
/kind bug

**What this PR does / why we need it**:
Updates v1.Taint parser to accept the form `key:effect`, which is the [string representation](https://github.com/kubernetes/kubernetes/blob/33b20f2/pkg/apis/core/taint.go#L33) of a v1.Taint that has a nil Value (the empty string).

Also updates the parser to accept the form `key=:effect-`, since `key=:effect` is a valid string representation of a v1.Taint with a nil Value.

- [x] Update code and tests in `pkg/util/taints/`
- [x] Update [kubectl's copy of the code and tests](https://github.com/kubernetes/kubernetes/pull/69446) in `pkg/kubectl/cmd/taint/`
- [x] Update `kubectl taint` usage
 
**Which issue(s) this PR fixes**:
Fixes #73249

**Does this PR introduce a user-facing change?**:
```release-note
Support parsing more v1.Taint forms. `key:effect`, `key=:effect-` are now accepted.
```

/cc @liggitt

### CODE REVIEW

**Review by neolit123 — COMMENTED** [review]
thanks @dlipovetsky 
/priority important-longterm

could you please add a user facing release note instead of NONE? e.g.
> `Support parsing v1.Taint forms such as xx, yy`

**Comment by dlipovetsky** [comment]
Thanks @neolit123! Added release note.

**Comment by dlipovetsky** [comment]
I added two new invalid spec tests and fixed a bug in the updated parser. (I also verified that the original parser passed these new tests)

**Comment by dlipovetsky** [comment]
Once the changes to `pkg/util/taints/` are finalized, I'll make corresponding changes to `pkg/kubectl/cmd/taint/`.

**Review by Huang-Wei — COMMENTED** [review]
LGTM for the pkg/util/taints/ part.

**Comment by kevin-wangzefeng** [comment]
LGTM for `pkg/util/taints/`, I think you can go ahead for the rest parts.

**Comment by dlipovetsky** [comment]
Thanks @kevin-wangzefeng and @Huang-Wei.

After I updated `parseTaints` and `parseTaint` in `pkg/kubectl/cmd/taint/util.go`, an [existing test](https://github.com/kubernetes/kubernetes/blob/f9e102cac7b08551d978764e3ac62e68fc6a2666/pkg/kubectl/cmd/taint/taint_test.go#L164) in `pkg/kubectl/cmd/taint/taint_test.go` failed:
```
--- FAIL: TestTaint (0.03s)
    --- FAIL: TestTaint/node_has_two_taints_with_the_same_key_but_different_effect,_remove_all_of_them_with_wildcard (0.00s)
        taint_test.go:313: Recovered: error: invalid taint spec: dedicated
            See 'taint -h' for help and examples
        taint_test.go:332: node has two taints with the same key but different effect, remove all of them with wildcard: node not tainted
```
I did not realize that `kubectl taint` accepts _a key with no value or effect_ to remove a taint; the example in the kubectl usage is: 
```
  # Remove from node 'foo' all the taints with key 'dedicated'
  kubectl taint nodes foo dedicated-
``` 
This use case was introduced in https://github.com/kubernetes/kubernetes/pull/30590. To be backward compatible, this PR should not break this use case.

There was no test in `pkg/util/taints` for removing a taint using the `key-` form. The parser prior to this PR accepts this form, but the parser in this PR rejects this form. My plan is to:
1. Add a test to `pkg/util/taints` for removing a taint using the `key-` form
2. Update the parser in this PR to accept the `key-` form

**Comment by dlipovetsky** [comment]
The last commit passes the tests in `pkg/util/taints` and `pkg/kubectl/cmd/taint/`.

I have changed `parseTaint` to accept the `key` form, and ensured in `ParseTaints` that a taint with no effect cannot be added. The alternative is for `parseTaint` to continue rejecting this form, and handling the form as a special case in `ParseTaints`. I prefer the former, because the code already passes around taints without values or effects, which implies that the `key` form is valid, although it must not be used to create a taint.

**Comment by kevin-wangzefeng** [comment]
I'm fine with it. Empty/null value and effect should not block parsing, though taints with only keys indicated should not be used for creating.

**Comment by dlipovetsky** [comment]
Thanks for the feedback @kevin-wangzefeng.

I've copied the changes to `pkg/kubectl/cmd/taint` and updated the kubectl taint usage.

**Comment by jayunit100** [comment]
@derekwaynecarr @aveshagarwal @ravisantoshgudimetla ~ can one of you help us push this taints patch through ? would like to possibly add some more related patches in the near future as well. (needs /approve) :)

**Comment by Huang-Wei** [comment]
@dlipovetsky can you squash the PR to 3 commits? - each commit mapped to each task.

And only elders in pkg/OWNERS can approve changes in `pkg/util/taints/` :)

**Comment by dlipovetsky** [comment]
@Huang-Wei Good idea, will do. Didn't want to squash too early.

Should I choose someone from pkg/OWNERS at random? :smile:

File: pkg/kubectl/cmd/taint/utils.go
Comment (bsalamat) [inline pkg/kubectl/cmd/taint/utils.go]: Why `<key>` is considered a valid taint here? It looks like it gets rejected later in `parseTaints` when adding a taint, but it is accepted when removing a taint. If this behavior is desired, it should be documented in the comments and the release note.

**Review by bsalamat — COMMENTED** [review]
Looks good and thanks for cleaning up parts of the logic, @dlipovetsky!
Just a few minor comments.

File: pkg/kubectl/cmd/taint/utils.go
Comment (bsalamat) [reply pkg/kubectl/cmd/taint/utils.go]: I just read [your comment](https://github.com/kubernetes/kubernetes/pull/74159#issuecomment-466575717) above. It is fine, but please update comments and the release note to point it out.

File: pkg/util/taints/taints.go
Comment (bsalamat) [inline pkg/util/taints/taints.go]: Please use `len(newTaint.Effect) == 0` here as well.

**Review by ravisantoshgudimetla — APPROVED** [review]
/lgtm

Thanks for the PR @dlipovetsky 

/hold

I will wait for @bsalamat and @liggitt comments but feel free to remove hold once it they're fine.

**Comment by liggitt** [comment]
looks reasonable at a quick glance, but I'll defer to @bsalamat for review

**Comment by bsalamat** [comment]
I reviewed and it looks fine. I have only a few minor comments. If @dlipovetsky addresses them sooner, we can get it merged.

File: pkg/kubectl/cmd/taint/taint.go
Comment (dlipovetsky) [reply pkg/kubectl/cmd/taint/taint.go]: Thanks. Fixed.

File: pkg/kubectl/cmd/taint/utils.go
Comment (dlipovetsky) [reply pkg/kubectl/cmd/taint/utils.go]: Fixed.

File: pkg/util/taints/taints.go
Comment (dlipovetsky) [reply pkg/util/taints/taints.go]: Thanks, fixed.

File: pkg/kubectl/cmd/taint/taint.go
Comment (bsalamat) [reply pkg/kubectl/cmd/taint/taint.go]: Sure. Please fix the example.

File: pkg/kubectl/cmd/taint/utils.go
Comment (bsalamat) [reply pkg/kubectl/cmd/taint/utils.go]: Did you push your changes?

File: pkg/kubectl/cmd/taint/utils.go
Comment (dlipovetsky) [reply pkg/kubectl/cmd/taint/utils.go]: I updated the comment to `ParseTaints`, since that's the exported method.

I do want to point out that the exported behavior has not changed. Prior to this PR, the `ParseTaints` function parsed some forms (including  `<key>`) itself, but delegated parsing of other forms to `parseTaint`. This PR moves all parsing to `parseTaint`.

Given this, do you think updating the release note is necessary?

File: pkg/kubectl/cmd/taint/utils.go
Comment (dlipovetsky) [reply pkg/kubectl/cmd/taint/utils.go]: My apologies, I didn't! I have pushed them now.

**Comment by dlipovetsky** [comment]
@bsalamat Thanks a lot for your review. My apologies for taking a while to address your comments. I address them using separate commits for the code in `pkg/util/taints` and `pkg/kubectl/cmd/taint`, so that I can easily fixup the original commits for those packages before merging.

**Comment by dlipovetsky** [comment]
I also added a test to verify that the empty string is an invalid Taint spec.

File: pkg/kubectl/cmd/taint/taint.go
Comment (dlipovetsky) [reply pkg/kubectl/cmd/taint/taint.go]: Fixed!

**Comment by dlipovetsky** [comment]
Test failed due to known flake, see #74931

/retest

**Comment by bsalamat** [comment]
Thanks, @dlipovetsky! Looks good. Could you please squash commits?

/lgtm

**Comment by bsalamat** [comment]
I updated the release note. Please check it.

**Comment by dlipovetsky** [comment]
@bsalamat Squashed.

As for the release note, I'm not sure it needs to mention the `key-` form, since that has always been accepted--this PR just changes where its parsed internally (please see https://github.com/kubernetes/kubernetes/pull/74159#discussion_r264408700).

**Comment by bsalamat** [comment]
@dlipovetsky Thanks again. I fixed the release notes.

/lgtm
/approve

**Comment by bsalamat** [comment]
@liggitt Back to you for approval.

**Comment by dlipovetsky** [comment]
/assign @liggitt 

(So that this PR shows up in your dashboard :smile:)

**Comment by dlipovetsky** [comment]
@ravisantoshgudimetla Please cancel the hold. Thanks!

**Pull request merged**
